Examples/c++/Linked List/DLList.h

```diff
@@ -0,0 +1,401 @@
+#pragma once
+
+#include <algorithm>
+#include <utility>
+
+#ifdef DEBUG
+	#define ASSERT(x,...) { if(!(x)){ std::cout << __VA_ARGS__ << std::endl; __debugbreak(); } }
+#else
+	#define ASSERT(x,...)
+#endif
+
+namespace DSA
+{
+	template<typename List>
+	class DLLIterator
+	{
+	public:
+		using ValueType = typename List::ValueType;
+		using NodePtr = typename List::NodePtr;
+		using PointerType = ValueType*;
+		using RefType = ValueType&;
+	public:
+		DLLIterator()
+			: m_Ptr(nullptr), m_SentinelNode(nullptr)
+		{
+		}
+
+		DLLIterator(NodePtr ptr, NodePtr sentinelNode)
+			: m_Ptr(ptr), m_SentinelNode(sentinelNode)
+		{
+		}
+
+		DLLIterator& Increment(int32_t offset)
+		{
+			ASSERT(offset >= 0, "Invalid offset!");
+
+			for (int32_t i = 0; i < offset; i++)
+				(*this)++;
+
+			return *this;
+		}
+
+		DLLIterator& Decrement(int32_t offset)
+		{
+			ASSERT(offset >= 0, "Invalid offset!");
+
+			for (int32_t i = 0; i < offset; i++)
+				(*this)--;
+
+			return *this;
+		}
+
+		RefType operator*()
+		{
+			ASSERT(m_Ptr != m_SentinelNode, "Cannot dereference end iterator!");
+			return m_Ptr->Data;
+		}
+
+		NodePtr operator->()
+		{
+			ASSERT(m_Ptr != m_SentinelNode, "Cannot access end interator!");
+			return m_Ptr;
+		}
+
+		DLLIterator& operator++()
+		{
+			ASSERT(m_Ptr != m_SentinelNode, "Cannot increment end iterator!");
+			m_Ptr = m_Ptr->Next;
+			return *this;
+		}
+
+		DLLIterator operator++(int)
+		{
+			DLLIterator<List> iterator = *this;
+			++(*this);
+			return iterator;
+		}
+
+		DLLIterator& operator--()
+		{
+			ASSERT(m_Ptr->Previous != m_SentinelNode, "Cannot decrement begin iterator!");
+			m_Ptr = m_Ptr->Previous;
+			return *this;
+		}
+
+		DLLIterator operator--(int)
+		{
+			DLLIterator<List> iterator = *this;
+			--(*this);
+			return iterator;
+		}
+
+		bool operator==(const DLLIterator& rhs) const
+		{
+			return m_Ptr == rhs.m_Ptr;
+		}
+
+		bool operator!=(const DLLIterator& rhs) const
+		{
+			return !(*this == rhs);
+		}
+
+		operator NodePtr()
+		{
+			return m_Ptr;
+		}
+	private:
+		NodePtr m_Ptr;
+		NodePtr m_SentinelNode;
+	};
+
+
+	template<typename T>
+	struct DLNode
+	{
+		using NodePtr = DLNode<T>*;
+
+		T Data = T();
+		NodePtr Next = nullptr;
+		NodePtr Previous = nullptr;
+
+		DLNode() = default;
+		DLNode(const T& data, NodePtr prev, NodePtr next)
+			: Data(data), Previous(prev), Next(next) {}
+		DLNode(T&& data, NodePtr prev, NodePtr next)
+			: Data(std::move(data)), Previous(prev), Next(next) {}
+	};
+
+	template<typename T>
+	class DLList
+	{
+	public:
+		using ValueType = T;
```

Comment:
are this usings really needed, I dont see them actually improving the readibility.

Reply:
Some of the usings are indeed not used and will be removed from the code.